### PR TITLE
Numba integration

### PR DESCRIPTION
**Pull request by Emerald01**
the full support of Numba
1. CUDADataManager and CUDAFunctionManager and other CUDAXXXX classes as the base class
2. PyCUDAXXXX and NumbaXXXX as derived classes
3. common APIs for both pycuda and numba in EnvWrapper, with `env_backend = cpu, pycuda or numba`. APIs are also compatible with the version <= 1.7
4. trainers for both backend modes
5. numba core services including sampler, resetter and all others for end-to-end simulation and train
6. numba env of Tag Continuous, implemented by Brenton Chu @blchu 
7. unittests of all codebase for both backend modes and same testings

### CODE REVIEW

File: warp_drive/utils/argument_fix.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import functools
+import warnings
+
+
+class Argfix:
+    """Decorator ensuring backward compatibility when an argument name is
```
Comment (StephanZheng) [inline warp_drive/utils/argument_fix.py:6]: Can you clarify when we modify argument names? That's kinda dangerous...

File: tests/warp_drive/numba_codes/__init__.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2021, salesforce.com, inc.
```
Comment (StephanZheng) [inline tests/warp_drive/numba_codes/__init__.py:1]: Should we rename the top folder "numba_codes" -> "numba_tests"?

File: tests/warp_drive/pycuda_codes/__init__.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2021, salesforce.com, inc.
```
Comment (StephanZheng) [inline tests/warp_drive/pycuda_codes/__init__.py:1]: Should we rename the top folder "pycuda_codes" -> "pycuda_tests"?

File: tests/multiblocks_per_env/__init__.py [hunk 4]
```diff
@@ -3,3 +3,4 @@
 # SPDX-License-Identifier: BSD-3-Clause
```
Comment (StephanZheng) [inline tests/multiblocks_per_env/__init__.py:3]: Should we rename the top folder "mulitblocks_per_env" -> "mulitple_cuda_blocks_per_env"?

**Review by StephanZheng — CHANGES_REQUESTED** [review]
I'd rename every instances "XYZ_codes" to just "XYZ"

File: warp_drive/utils/argument_fix.py (hunk 1 above)
Comment (Emerald01) [reply warp_drive/utils/argument_fix.py:6]: added comments

**Pull request merged**